OpenSim/Common/Component.cpp

```diff
@@ -838,26 +836,74 @@ void Component::constructOutputForStateVariable(const std::string& name)
             SimTK::Stage::Model);
 }
 
+// mark components owned as properties as subcomponents
+void Component::markPropertiesAsSubcomponents()
+{
+    // if being invoked either constructing a new Component
+    // or the properties have been modified. In the latter case
+    // we must make sure that pointers to old properties are cleared
+    _components.clear();
+
+    // Now mark properties that are Components as subcomponents
+    //loop over all its properties
+    for (int i = 0; i < getNumProperties(); ++i) {
+        auto& prop = updPropertyByIndex(i);
+        // check if property is of type Object
+        if (prop.isObjectProperty()) {
+            // a property is a list so cycle through its contents
+            for (int j = 0; j < prop.size(); ++j) {
+                Object& obj = prop.updValueAsObject(j);
+                // if the object is a Component mark it
+                if (Component* comp = dynamic_cast<Component*>(&obj) ) {
+                    markAsSubcomponent(comp);
+                }
+                else {
+                    // otherwise it may be a Set (of objects), and
+                    // would prefer to do something like this to test:
+                    //  Set<Component>* objects = dynamic_cast<Set<Component>*>(obj)
+                    // Instead we can see if the object has a property called
+                    // "objects" which is a PropertyObjArray used by Set<T>.
+                    // knowing the object Type is useful for debugging
+                    // and it could be used to strengthen the test (e.g. scan  
+                    // for "Set" in the type name). 
+                    std::string objType = obj.getConcreteClassName();
+                    if (obj.hasProperty("objects")) {
+                        // get the PropertyObjArray if the object has one
+                        auto& objectsProp = obj.updPropertyByName("objects");
+                        // loop over the objects in the PropertyObjArray
+                        for (int k = 0; k < objectsProp.size(); ++k) {
+                            Object& obj = objectsProp.updValueAsObject(k);
+                            // if the object is a Component mark it
+                            if (Component* comp = dynamic_cast<Component*>(&obj) )
+                                markAsSubcomponent(comp);
+                        } // loop over objects and mark it if it is a component
+                    } // end if property is a Set with "objects" inside
+                } // end of if/else property value is an Object or something else
+            } // loop over the property list
+        } // end if property is an Object
+    } // loop over properties
+}
+
 // Include another Component as a subcomponent of this one. If already a
 // subcomponent, it is not added to the list again.
-void Component::addComponent(Component* component)
+void Component::markAsSubcomponent(Component* component)
```

Comment:
The name of this method and the fact that `component` is passed in writable makes me think that it is `component` that will be modified. Could this method be something like `recordInSubcomponentList()`?

Also it would be good to use the term `subcomponent` whenever it can help distinguish parent from child, so I think the argument here should be `Component& subcomponent`.